src/site/content/en/blog/webassembly-memory-debugging/index.md

```diff
@@ -0,0 +1,607 @@
+C++ codec wrappers. Let's take a look at an [ImageQuant](https://pngquant.org/lib/) wrapper as an
+example (reduced to show only object creation and deallocation parts):
+
+```cpp
+liq_attr* attr;
+liq_image* image;
+liq_result* res;
+uint8_t* result;
+
+RawImage quantize(std::string rawimage,
+                  int image_width,
+                  int image_height,
+                  int num_colors,
+                  float dithering) {
+  const uint8_t* image_buffer = (uint8_t*)rawimage.c_str();
+  int size = image_width * image_height;
+
+  attr = liq_attr_create();
+  image = liq_image_create_rgba(attr, image_buffer, image_width, image_height, 0);
+  liq_set_max_colors(attr, num_colors);
+  liq_image_quantize(image, attr, &res);
+  liq_set_dithering_level(res, dithering);
+  uint8_t* image8bit = (uint8_t*)malloc(size);
+  result = (uint8_t*)malloc(size * 4);
+
+  // …
+
+  free(image8bit);
+  liq_result_destroy(res);
+  liq_image_destroy(image);
+  liq_attr_destroy(attr);
+
+  return {
+    val(typed_memory_view(image_width * image_height * 4, result)),
+    image_width,
+    image_height
+  };
+}
+
+void free_result() {
+  free(result);
+}
+```
+
+JavaScript (well, TypeScript):
+
+```ts
+export async function process(data: ImageData, opts: QuantizeOptions) {
+  if (!emscriptenModule) {
+    emscriptenModule = initEmscriptenModule(imagequant, wasmUrl);
+  }
+  const module = await emscriptenModule;
+
+  const result = module.quantize(/* … */);
+
+  module.free_result();
+
+  return new ImageData(
+    new Uint8ClampedArray(result.view),
+    result.width,
+    result.height
+  );
+}
+```
+
+Do you spot an issue? Hint: it's
+[use-after-free](https://owasp.org/www-community/vulnerabilities/Using_freed_memory), but in
+JavaScript!
+
+In Emscripten, `typed_memory_view` returns a JavaScript `Uint8Array` backed by the WebAssembly
+memory buffer, with `byteOffset` and `byteLength` set to the given pointer and length. The main
+point is that this is a TypedArray *view* into a WebAssembly memory buffer, rather than a
+JavaScript-owned copy of the data.
+
+When we call `free_result` from JavaScript, it, in turn, calls a standard C function `free` to mark
+this memory as available for any future allocations, which means the data that our `Uint8Array` view
+points to, can be overwritten with arbitrary data by any future call into Wasm.
+
+Or, some implementation of `free` might even decide to zero-fill the freed memory immediately. The
+`free` that Emscripten uses doesn't do that, but we are relying on an implementation detail here
+that cannot be guaranteed.
+
+Or, even if the memory behind the pointer gets preserved, new allocation might need to grow the
+WebAssembly memory. When `WebAssembly.Memory` is grown either via JavaScript API, or corresponding
+`memory.grow` instruction, it invalidates the existing `ArrayBuffer` and, transitively, any views
+backed by it.
+
+Let me use DevTools or Node.js console to demonstrate this behaviour:
```

Comment:
```suggestion
Let me use the DevTools (or Node.js) console to demonstrate this behavior:
```

The sentence looks a bit grammatically off without `the`

Style guide tells us to use American spelling: https://developers.google.com/style/spelling